lib/modules/apostrophe-attachments/lib/api.js

```diff
@@ -944,13 +959,29 @@ module.exports = function(self, options) {
     ], callback);
 
     function updateCounts(callback) {
-      return async.eachSeries(commands, function(command, callback) {
-        return self.db.update(command[0], command[1], callback);
+      return async.eachSeries(commands, function([criteria, query], callback) {
+        return self.db.update(criteria, query, callback);
       }, callback);
     }
 
   };
 
+  // We check that no attachment is possible for a doc by
+  // parsing its schema recursively until we find an 'attachment' type.
+  // If a doc has currently no attachments but could have had some in the past, it keeps updating.
+  self.findAttachmentsInSchema = function(schema) {
+    if (!schema) {
+      return false;
+    }
+
+    return findAttachment(schema);
+
+    function findAttachment(fields) {
+      return fields.some((field) => field.type === 'attachment' ||
+        (field.schema && findAttachment(field.schema)));
```

Comment:
This works for arrays, but it does not work for widgets which are part of the document too.

If field.type is array or singleton, then you need to iterate the allowed widget types in it, and for each of those, call `self.apos.areas.getManager(widgetTypeName)` to get a manager. Then you can access `.schema` again and make a recursive call to your method.

Reply:
Thanks a lot, this is the piece I missed.